### PR TITLE
8291336: Add ideal rule to convert floating point multiply by 2 into addition

### PR DESCRIPTION
**Pull request by SuperCoder7979**
Hello,
I would like to propose an ideal transform that converts floating point multiply by 2 (`x * 2`) into an addition operation instead. This would allow for the elimination of the memory reference for the constant two, and keep the whole operation inside registers. My justifications for this optimization include:
* As per [Agner Fog's instruction tables](https://www.agner.org/optimize/instruction_tables.pdf) many older systems, such as the sandy bridge and ivy bridge architectures, have different latencies for addition and multiplication meaning this change could have beneficial effects when in hot code.
* The removal of the memory load would have a beneficial effect in cache bound situations.
* Multiplication by 2 is relatively common construct so this change can apply to a wide range of Java code.

As this is my first time looking into the c2 codebase, I have a few lingering questions about my implementation and how certain parts of the compiler work. Mainly, is this patch getting the type of the operands correctly? I saw some cases where code used `bottom_type()` and other cases where it used `phase->type(value)`. Similarly, are nodes able to be reused as is being done in the AddNode constructors? I saw some places where the clone method was being used, but other places where it wasn't.

I have attached an IR test and a jmh benchmark. Tier 1 testing passes on my machine.

Thanks for your time,
Jasmine

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8291336](https://bugs.openjdk.org/browse/JDK-8291336): Add ideal rule to convert floating point multiply by 2 into addition

### Reviewers
 * [Quan Anh Mai](https://openjdk.org/census#qamai) (@merykitty - Committer) ⚠️ Review applies to [d4303fad](https://git.openjdk.org/jdk/pull/9642/files/d4303fad11b78a0ae348c62159d8c65c92dd12c1)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9642/head:pull/9642` \
`$ git checkout pull/9642`

Update a local copy of the PR: \
`$ git checkout pull/9642` \
`$ git pull https://git.openjdk.org/jdk pull/9642/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9642`

View PR using the GUI difftool: \
`$ git pr show -t 9642`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9642.diff">https://git.openjdk.org/jdk/pull/9642.diff</a>

</details>

### CODE REVIEW

**Comment by SuperCoder7979** [comment]
Unfortunately it seems I can't open bugs on the JBS, is there a way to do so or will someone else have to do it for me?

**Comment by merykitty** [comment]
There you go: https://bugs.openjdk.org/browse/JDK-8291336

Next time you could ask for help in the appropriate mailing list (this time it is hotspot-compiler-dev) or submit a bug through https://bugreport.java.com/bugreport/

Also please enable github action in your fork so that the patches get tested automatically at tier 1 on major platforms.

Hope this helps.

File: test/hotspot/jtreg/compiler/c2/irTests/TestMulBy2.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+import jdk.test.lib.Utils;
+
+/*
+ * @test
+ * @summary Test that transformation of multiply-by-2 is appropriately turned into additions.
+ * @library /test/lib /
+ * @requires vm.compiler2.enabled
+ * @run driver compiler.c2.irTests.TestMulBy2
+ */
+public class TestMulBy2 {
```
Comment (merykitty) [inline test/hotspot/jtreg/compiler/c2/irTests/TestMulBy2.java:37]: Please use a more general name such as `MulFNodeIdealizationTests`

**Review by merykitty — COMMENTED** [review]
I leave some reviews for the patch, can you show the results of the added microbenchmark, please?

Thanks

**Comment by SuperCoder7979** [comment]
Hi, thank you for your assistance with this, I have updated the PR title and have applied the changes from code review. I have also updated the benchmark and have attached the results below. I tested the benchmark on 2 systems, a new one and an old one. The new system has a Ryzen 5 4500U cpu, and the results are as shown:
```
                                       Baseline                  Patch
Benchmark                 Mode  Cnt    Score   Error  Units      Score   Error  Units
TestMul2.testMul2Double   avgt   10  209.740 ± 1.454  ns/op // 209.315 ± 1.116  ns/op (+0.20%)
TestMul2.testMul2Float    avgt   10  210.871 ± 6.179  ns/op // 209.498 ± 0.777  ns/op (+0.65%)
```
The benchmark showed very little change on the new system, which is expected as the documentation states that both the `vaddsd` and `vmulsd` instructions have a latency of 3 cycles and a reciprocal throughput of 0.5. The slight gain could be from the elimination of the memory reference, or just from testing variance. The older system ran a Xeon x5690, and had these results:
```
                                       Baseline                  Patch
Benchmark                 Mode  Cnt    Score   Error  Units      Score   Error  Units
TestMul2.testMul2Double   avgt   10  190.062 ± 9.695  ns/op // 170.393 ± 1.193  ns/op (+10.34%)
TestMul2.testMul2Float    avgt   10  184.239 ± 1.983  ns/op // 171.329 ± 4.261  ns/op (+7.00%)
```
Due to the older system having a faster addition than multiplication, especially with double precision operations, far more substantial gains were realized here.

File: src/hotspot/share/opto/mulnode.cpp [hunk 2]
```diff
@@ -433,6 +433,31 @@ const Type *MulFNode::mul_ring(const Type *t0, const Type *t1) const {
   return TypeF::make( t0->getf() * t1->getf() );
 }
 
+//------------------------------Ideal---------------------------------------
+// Check to see if we are multiplying by a constant 2 and convert to add, then try the regular MulNode::Ideal
+Node *MulFNode::Ideal(PhaseGVN *phase, bool can_reshape) {
+  const TypeF *t1 = in(1)->bottom_type()->isa_float_constant();
```
[the remark this answers is not in the document]
Comment (SuperCoder7979) [reply src/hotspot/share/opto/mulnode.cpp]: Thank you for this perspective! I have updated the code to use `phase->type()`.

File: src/hotspot/share/opto/mulnode.cpp [hunk 3]
```diff
@@ -433,6 +433,31 @@ const Type *MulFNode::mul_ring(const Type *t0, const Type *t1) const {
...
+  const TypeF *t2 = in(2)->bottom_type()->isa_float_constant();
+
+  // x * 2 -> x + x
```
[the remark this answers is not in the document]
Comment (SuperCoder7979) [reply src/hotspot/share/opto/mulnode.cpp:441]: Done, thanks!

File: test/hotspot/jtreg/compiler/c2/irTests/TestMulBy2.java (hunk 1 above)
Comment (SuperCoder7979) [reply test/hotspot/jtreg/compiler/c2/irTests/TestMulBy2.java]: Done

**Review by TobiHartmann — CHANGES_REQUESTED** [review]
Looks good to me otherwise. I executed some testing and it all passed.

File: src/hotspot/share/opto/mulnode.cpp [hunk 4]
```diff
@@ -433,6 +433,22 @@ const Type *MulFNode::mul_ring(const Type *t0, const Type *t1) const {
...
 }
 
+//------------------------------Ideal---------------------------------------
+// Check to see if we are multiplying by a constant 2 and convert to add, then try the regular MulNode::Ideal
+Node *MulFNode::Ideal(PhaseGVN *phase, bool can_reshape) {
+  const TypeF *t2 = phase->type(in(2))->isa_float_constant();
+
+  // x * 2 -> x + x
+  if (t2 != NULL) {
+    if (t2->getf() == 2) {
```
[the remark this answers is not in the document]
Comment (SuperCoder7979) [reply src/hotspot/share/opto/mulnode.cpp]: Completed, thanks!

**Comment by SuperCoder7979** [comment]
Thank you for testing! I have applied changes from code review.

**Review by TobiHartmann — APPROVED** [review]
Looks reasonable to me but a second review would be good.

**Comment by SuperCoder7979** [comment]
Hi, may I have more reviews on this change, please?
Thanks

File: src/hotspot/share/opto/mulnode.cpp [hunk 5]
```diff
@@ -433,6 +433,20 @@ const Type *MulFNode::mul_ring(const Type *t0, const Type *t1) const {
   return TypeF::make( t0->getf() * t1->getf() );
 }
 
+//------------------------------Ideal---------------------------------------
+// Check to see if we are multiplying by a constant 2 and convert to add, then try the regular MulNode::Ideal
+Node *MulFNode::Ideal(PhaseGVN *phase, bool can_reshape) {
```
Comment (merykitty) [inline src/hotspot/share/opto/mulnode.cpp:438]: Please use the format `Type* identifier` for new code (in this case it is `PhaseGVN* phase`). The same applies to other places.

File: test/hotspot/jtreg/compiler/c2/irTests/TestMulNodeIdealization.java [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    @Test
+    @IR(failOn = {IRNode.MUL})
+    // Checks x * 2 -> x + x
+    public double testDouble(double x) {
+        return x * 2;
+    }
+
+    @Run(test = "testFloat")
+    public void runTestFloat() {
+        testFloat(RANDOM.nextFloat());
```
Comment (merykitty) [inline test/hotspot/jtreg/compiler/c2/irTests/TestMulNodeIdealization.java:61]: Verification against the execution in the interpreter would be better here.

**Review by merykitty — APPROVED** [review]
The change looks good to me otherwise. Thanks.

**Comment by SuperCoder7979** [comment]
Hi, apologies for the delayed reply but I have fixed the style and have added verification of the optimization against the interpreter. A re-review would be much appreciated. Thanks for your time once again!

File: test/hotspot/jtreg/compiler/c2/irTests/TestMulNodeIdealization.java (hunk 6 above)
Comment (SuperCoder7979) [reply test/hotspot/jtreg/compiler/c2/irTests/TestMulNodeIdealization.java]: Completed, thanks!

File: src/hotspot/share/opto/mulnode.cpp (hunk 5 above)
Comment (SuperCoder7979) [reply src/hotspot/share/opto/mulnode.cpp]: Done

**Review by TobiHartmann — APPROVED** [review]
Still looks good to me.

**Comment by SuperCoder7979** [comment]
Thank you!

**Review by chhagedorn — APPROVED** [review]
Looks good! I quickly run some sanity testing again and then sponsor it.

**Comment by chhagedorn** [comment]
Testing looked good!

/sponsor

**Pull request closed without merging**